pkg/tfbridge/provider_test.go

```diff
@@ -1152,6 +1156,148 @@ func TestCheck(t *testing.T) {
+									Required: true,
+									Elem:     &schema.Schema{Type: schema.TypeString},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	// we need the pschema for type checking
+	pulumiSchemaSpec := &pschema.PackageSpec{
+		Resources: map[string]pschema.ResourceSpec{
+			"ExampleResource": {
+				StateInputs: &pschema.ObjectTypeSpec{
+					Properties: map[string]pschema.PropertySpec{
+						"networkConfiguration": {
+							TypeSpec: pschema.TypeSpec{
+								Ref: "#/types/testprov:ExampleResourceNetworkConfiguration",
+							},
+						},
+					},
+				},
+				InputProperties: map[string]pschema.PropertySpec{
+					"networkConfiguration": {
+						TypeSpec: pschema.TypeSpec{
+							Ref: "#/types/testprov:ExampleResourceNetworkConfiguration",
+						},
+					},
+				},
+				ObjectTypeSpec: pschema.ObjectTypeSpec{
+					Properties: map[string]pschema.PropertySpec{
+						"networkConfiguration": {
+							TypeSpec: pschema.TypeSpec{
+								Ref: "#/types/testprov:ExampleResourceNetworkConfiguration",
+							},
+						},
+					},
+				},
+			},
+		},
+		Types: map[string]pschema.ComplexTypeSpec{
+			"testprov:ExampleResourceNetworkConfiguration": {
+				ObjectTypeSpec: pschema.ObjectTypeSpec{
+					Properties: map[string]pschema.PropertySpec{
+						"securityGroups": {
+							TypeSpec: pschema.TypeSpec{
+								Type: "array",
+								Items: &pschema.TypeSpec{
+									Type: "string",
+								},
+							},
+						},
+						"subnets": {
+							TypeSpec: pschema.TypeSpec{
+								Type: "array",
+								Items: &pschema.TypeSpec{
+									Type: "string",
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	provider := &Provider{
+		tf:               shimv2.NewProvider(p, shimv2.WithDiffStrategy(shimv2.PlanState)),
+		module:           "testprov",
+		config:           shimv2.NewSchemaMap(p.Schema),
+		pulumiSchema:     []byte("hello"), // we only check whether this is nil in type checking
+		pulumiSchemaSpec: pulumiSchemaSpec,
+		resources: map[tokens.Type]Resource{
+			"ExampleResource": {
+				TF:     shimv2.NewResource(p.ResourcesMap["example_resource"]),
+				TFName: "example_resource",
+				Schema: &ResourceInfo{
+					Tok: "ExampleResource",
+				},
+			},
+		},
+	}
+
+	args, err := structpb.NewStruct(map[string]interface{}{
+		"networkConfiguration": []interface{}{
+			map[string]interface{}{
+				"securityGroups": []interface{}{
+					"04da6b54-80e4-46f7-96ec-b56ff0331ba9",
+				},
+				"subnets": "[\"first\",\"second\"]", // this is a type error
+			},
+		},
+	})
+	require.NoError(t, err)
+	_, err = provider.Check(ctx, &pulumirpc.CheckRequest{
+		Urn:  "urn:pulumi:dev::teststack::ExampleResource::exres",
+		Olds: &structpb.Struct{},
+		News: args,
+	})
+	require.NoError(t, err)
+
+	// run 'go test  -run=TestCheckWarnings -v ./pkg/tfbridge/ -update' to update
+	autogold.Expect(`warning: Type checking failed:
+warning: Unexpected type at field "networkConfiguration":
```

Comment:
Each time you call `logger.Warn()` it prints out `warning: `. Personally I don't mind this.